src/lib/transition_router/initial_validator.ml

```diff
@@ -179,52 +179,56 @@ let run ~logger ~trust_system ~verifier ~transition_reader
   let duplicate_checker = Duplicate_block_detector.create () in
   don't_wait_for
     (Reader.iter transition_reader ~f:(fun network_transition ->
-         if Ivar.is_full initialization_finish_signal then (
+         if Ivar.is_full initialization_finish_signal then
            let ( `Transition transition_env
                , `Time_received time_received
-               , `Valid_cb is_valid_cb ) =
+               , `Valid_cb valid_cb ) =
              network_transition
            in
-           let transition_with_hash =
-             Envelope.Incoming.data transition_env
-             |> With_hash.of_data
-                  ~hash_data:
-                    (Fn.compose Protocol_state.hash
-                       External_transition.protocol_state)
-           in
-           Duplicate_block_detector.check ~precomputed_values duplicate_checker
-             logger transition_with_hash ;
-           let sender = Envelope.Incoming.sender transition_env in
-           let defer f = Fn.compose Deferred.return f in
-           match%bind
-             let open Deferred.Result.Monad_infix in
-             External_transition.(
-               Validation.wrap transition_with_hash
-               |> defer
-                    (validate_time_received ~precomputed_values ~time_received)
-               >>= defer (validate_genesis_protocol_state ~genesis_state_hash)
-               >>= (fun x -> validate_proofs ~verifier [x] >>| List.hd_exn)
-               >>= defer validate_delta_transition_chain
-               >>= defer validate_protocol_versions)
-           with
-           | Ok verified_transition ->
-               External_transition.poke_validation_callback
-                 (Envelope.Incoming.data transition_env)
-                 is_valid_cb ;
-               Envelope.Incoming.wrap ~data:verified_transition ~sender
-               |> Writer.write valid_transition_writer ;
-               let blockchain_length =
-                 External_transition.Initial_validated.consensus_state
-                   verified_transition
-                 |> Consensus.Data.Consensus_state.blockchain_length
-                 |> Coda_numbers.Length.to_int
-               in
-               Coda_metrics.Transition_frontier.update_max_blocklength_observed
-                 blockchain_length ;
-               return ()
-           | Error error ->
-               is_valid_cb `Reject ;
-               handle_validation_error ~logger ~trust_system ~sender
-                 ~state_hash:(With_hash.hash transition_with_hash)
-                 ~delta:genesis_constants.protocol.delta error )
+           if not (Coda_net2.Validation_callback.is_expired valid_cb) then (
```

Comment:
Should we be using an `Interruptible.t` for this, interrupted by the expiration signal, to avoid doing extra work if it expires while we are validating?

Reply:
That's a good idea, let me try and set that up.